src/python/grpcio_tests/tests/reflection/_reflection_client_test.py

```diff
@@ -0,0 +1,151 @@
+        self.assertEqual(file_name, file_desc.name)
+        self.assertEqual(_PROTO_PACKAGE_NAME, file_desc.package)
+        self.assertEqual("proto3", file_desc.syntax)
+        self.assertIn("TestService", file_desc.services_by_name)
+
+        file_name = _EMPTY_PROTO_FILE_NAME
+        file_desc = self.desc_pool.FindFileByName(file_name)
+        self.assertEqual(file_name, file_desc.name)
+        self.assertEqual(_PROTO_PACKAGE_NAME, file_desc.package)
+        self.assertEqual("proto3", file_desc.syntax)
+        self.assertIn("Empty", file_desc.message_types_by_name)
+
+    def testFindFileError(self):
+        with self.assertRaises(KeyError):
+            self.desc_pool.FindFileByName(_INVALID_FILE_NAME)
+
+    def testFindMessage(self):
+        message_name = _EMPTY_PROTO_SYMBOL_NAME
+        message_desc = self.desc_pool.FindMessageTypeByName(message_name)
+        self.assertEqual(message_name, message_desc.full_name)
+        self.assertTrue(message_name.endswith(message_desc.name))
+
+    def testFindMessageError(self):
+        with self.assertRaises(KeyError):
+            self.desc_pool.FindMessageTypeByName(_INVALID_SYMBOL_NAME)
+
+    def testFindServiceFindMethod(self):
+        service_name = self._SERVICE_NAMES[0]
+        service_desc = self.desc_pool.FindServiceByName(service_name)
+        self.assertEqual(service_name, service_desc.full_name)
+        self.assertTrue(service_name.endswith(service_desc.name))
+        file_name = _PROTO_FILE_NAME
+        file_desc = self.desc_pool.FindFileByName(file_name)
+        self.assertIs(file_desc, service_desc.file)
+
+        method_name = "EmptyCall"
+        self.assertIn(method_name, service_desc.methods_by_name)
+
+        method_desc = service_desc.FindMethodByName(method_name)
+        self.assertIs(method_desc, service_desc.methods_by_name[method_name])
+        self.assertIs(service_desc, method_desc.containing_service)
+        self.assertEqual(method_name, method_desc.name)
+        self.assertTrue(method_desc.full_name.endswith(method_name))
+
+        empty_message_desc = self.desc_pool.FindMessageTypeByName(
+            _EMPTY_PROTO_SYMBOL_NAME)
+        self.assertEqual(empty_message_desc, method_desc.input_type)
+        self.assertEqual(empty_message_desc, method_desc.output_type)
+
+    def testFindServiceError(self):
+        with self.assertRaises(KeyError):
+            self.desc_pool.FindServiceByName(_INVALID_SYMBOL_NAME)
+
+    def testFindMethodError(self):
+        service_name = self._SERVICE_NAMES[0]
+        service_desc = self.desc_pool.FindServiceByName(service_name)
+
+        # FindMethodByName sometimes raises a KeyError, and sometimes returns None.
+        # See https://github.com/protocolbuffers/protobuf/issues/9592
+        with self.assertRaises(KeyError):
+            res = service_desc.FindMethodByName(_INVALID_SYMBOL_NAME)
+            if res is None:
+                raise KeyError()
```

Comment:
Can you share more detail here? Are we going to remove this test, or what improvement did we add to make it compatible?

https://github.com/grpc/grpc/pull/29023#issuecomment-1060399713

Reply:
The original test was
```
        with self.assertRaises(KeyError):
            service_desc.FindMethodByName(_INVALID_SYMBOL_NAME)
```
I changed it to translate a `None` return value to a `raise KeyError`.

Reply:
Thanks for the clear write-up in https://github.com/protocolbuffers/protobuf/issues/9592.

Checking other find* methods ([Python code](https://github.com/protocolbuffers/protobuf/blob/349d74d92e8268c41fbb726b0a8292d588c63e20/python/google/protobuf/descriptor_database.py)), they favor the raise behavior. Do you know how the C-Extension implementation will behave? Based on [code](https://github.com/protocolbuffers/protobuf/blob/2d69d44cc18beaa0a3a62d353f80aead669abb5c/python/google/protobuf/pyext/descriptor_database.cc#L113), their return type is `bool`.

Reply:
In the code you linked, the return type is `bool` but internally it calls a Python function (`PyObject_CallMethod(...)`). In this context we have a call chain that look like `Python -> C++ -> Python`, where the outer Python code is the one doing the call `desc_database.FindFileByName(...)` and the inner is the `FindFileByName` implementation. I'm not an expert on this, but I believe that if the inner Python code raises an exception then it is immediately caught by the outer Python code - it doesn't even do exception folding of that C++ function, just sort of steals the context.

If the exception is not caught by the outer Python code, it looks like this:
```
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    desc_pool.FindFileByName("foo")
KeyError: "Couldn't find file foo"
```
Note how all the traceback of the inner `C++ -> Python` is missing as result of this unusual flow.

I hope this answers your question.